src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,169 @@
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+
+        if (index == size) {
+            add(value);
+            return;
+        }
+
+        Node<T> newNode = new Node<>(value);
+        Node<T> current = head;
+
+        if (index == 0) {
+            newNode.next = head;
+            head.prev = newNode;
+            head = newNode;
+        } else {
+            current = findNodeByIndex(index);
+            newNode.next = current;
+            newNode.prev = current.prev;
+            if (current.prev != null) {
+                current.prev.next = newNode;
+            }
+            current.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T item : list) {
+            add(item);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+
+        Node<T> current = findNodeByIndex(index);
+        return current.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = findNodeByIndex(index);
+
+        T oldValue = current.value;
+        current.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+
+        Node<T> current = head;
+        if (index == 0) {
+            head = head.next;
+            if (head != null) {
+                head.prev = null;
+            } else {
+                tail = null;
+            }
+        } else {
+            current = findNodeByIndex(index);
+            unlink(current);
+        }
+        size--;
+        return current.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+
+        while (current != null) {
+            if ((object == null && current.value == null)
+                    || (object != null && object.equals(current.value))) {
+                unlink(current);
+                size--;
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return (head == null);
+    }
+
+    private void unlink(Node<T> node) {
+        if (node == head) {
+            head = node.next;
+            return;
+        }
+        if (node == tail) {
+            tail = node.prev;
+            tail.next = null;
+            return;
+        }
+        node.prev.next = node.next;
+        node.next.prev = node.prev;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
```

Comment:
Please add more descriptive message, e.g '"Index " + index + " out of bounds. Valid range: 0 to " + (size - 1)"

Reply:
ok, done